proofs/tree.go

```diff
@@ -317,19 +311,50 @@ func (doctree *DocumentTree) AddLeavesFromDocument(document proto.Message) (err
 	if doctree.hash == nil {
 		return fmt.Errorf("hash is not set")
 	}
-	var getSalt GetSalt
-	if doctree.getSalt != nil {
-		getSalt = doctree.getSalt
+	var salts Salts
+	if doctree.salts != nil {
+		salts = doctree.salts
 	} else {
-		getSalt = defaultGetSalt(doctree.salts)
+		var err error
+		salts, err = defaultGetSalt(document)
+		if (err != nil){
+			return err
+		}
+	}
+	leaves, err := FlattenMessage(document, salts, doctree.readablePropertyLengthSuffix, doctree.hash, doctree.compactProperties, doctree.parentPrefix)
+	if err != nil {
+		return err
 	}
-	leaves, err := FlattenMessage(document, getSalt, doctree.saltsLengthSuffix, doctree.hash, doctree.compactProperties, doctree.parentPrefix)
+	err = fillBackSalts(document,leaves)
```

Comment:
The logic that is currently in fillBackSalts should be in the defaultSalts method.

Reply:
The reason why it should be in that method is that it should be customizable the same way defaultSalts is. It makes sense to put both in there I think. I've implemented this in the file that I linked to in the PR.